### PR TITLE
avoid deleting cbr0 address due to subnet string mismatch

### LINKED ISSUE #27873: KubeDNS is crashlooping after restarting kubelet on which it runs.
This is at least partially responsible for #27320 - some tests can't run when dns is down. 
Logs from just before the crash:
```
I0622 14:14:13.889398       1 logs.go:41] skydns: failure to forward request "read udp 10.180.1.8:43875->169.254.169.254:53: i/o timeout"
I0622 14:14:14.894338       1 dns.go:421] Received DNS Request:kubernetes.default.svc.cluster.local., exact:false
I0622 14:14:14.894365       1 dns.go:453] records:[], retval:[], path:[local cluster svc default kubernetes]
I0622 14:14:15.826099       1 dns.go:156] Ignoring error while waiting for service default/kubernetes: Get https://10.0.0.1:443/api/v1/namespaces/default/services/kubernetes: dial tcp 10.0.0.1:443: getsockopt: no route to host. Sleeping 1s before retrying.
E0622 14:14:15.826174       1 reflector.go:216] pkg/dns/dns.go:139: Failed to list *api.Service: Get https://10.0.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.0.0.1:443: getsockopt: no route to host
E0622 14:14:15.826216       1 reflector.go:216] pkg/dns/dns.go:138: Failed to list *api.Endpoints: Get https://10.0.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.0.0.1:443: getsockopt: no route to host
F0622 14:14:18.682159       1 server.go:124] Received signal: terminated
```

It got killed because Kubelet decided to kill it (readiness probe failed):
```
I0622 14:14:18.379547    8526 prober.go:106] Readiness probe for "kube-dns-v15-scynu_kube-system(c4a78332-3881-11e6-a37e-42010af00002):kubedns" failed (failure): Get http://10.180.1.8:8081/readiness: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
I0622 14:14:18.380077    8526 server.go:651] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"kube-dns-v15-scynu", UID:"c4a78332-3881-11e6-a37e-42010af00002", APIVersion:"v1", ResourceVersion:"277", FieldPath:"spec.containers{kubedns}"}): type: 'Warning' reason: 'Unhealthy' Readiness probe failed: Get http://10.180.1.8:8081/readiness: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
I0622 14:14:18.380134    8526 prober.go:106] Liveness probe for "kube-dns-v15-scynu_kube-system(c4a78332-3881-11e6-a37e-42010af00002):kubedns" failed (failure): Get http://10.180.1.8:8080/healthz: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
I0622 14:14:18.380162    8526 kubelet.go:2577] SyncLoop (container unhealthy): "kube-dns-v15-scynu_kube-system(c4a78332-3881-11e6-a37e-42010af00002)"
```

It started happening quite recently - I'm trying to locate a build which caused this. Repro:
run "Kubelet should not restart containers across restart" test - it caused this behavior for me multiple times today.

cc @wojtek-t @piosz @davidopp @kubernetes/goog-cluster - I applied the P0/v1.3 label/milestone, as I think it should be a blocker. Feel free to decrease the priority if you disagree @goltermann

### PR DESCRIPTION
**Pull request by freehan**
fix: #27873

### CODE REVIEW

File: pkg/util/net/util.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+*/
+
+package net
+
+import (
+	"net"
+	"reflect"
+)
+
+func IPNetEqual(ipnet1, ipnet2 *net.IPNet) bool {
```
Comment (dchen1107) [inline pkg/util/net/util.go]: nit: comment for future maintenance?

**Comment by dchen1107** [comment]
LGTM

File: pkg/util/net/util.go (hunk 1 above)
Comment (dchen1107) [inline pkg/util/net/util.go]: or unittests to cover the case?

**Comment by dchen1107** [comment]
Thanks for adding the unittests.

**Comment by girishkalele** [comment]
LGTM too - the unit tests help confirm the new logic.

**Comment by freehan** [comment]
I forgot to put the license header in unit test...

File: pkg/kubelet/network/kubenet/kubenet_linux.go [hunk 2]
```diff
@@ -249,8 +250,8 @@ func (plugin *kubenetNetworkPlugin) clearBridgeAddressesExcept(keep string) {
 	}
 
 	for _, addr := range addrs {
-		if addr.IPNet.String() != keep {
-			glog.V(5).Infof("Removing old address %s from %s", addr.IPNet.String(), BridgeName)
+		if !utilnet.IPNetEqual(addr.IPNet, keep) {
```
Comment (bprashanth) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: so the pod cidr changes? or is there something else going on here

Comment (freehan) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: It is just the string mismatch. 
`10.180.2.1/24` vs `10.180.2.0/24`

**Comment by bprashanth** [comment]
@kubernetes/sig-node consider adding kubelet restart to a blocking suite? 
It's currently part of Disruptive which is part of serial, which was probably disabled for this bug. Maybe we can just make serial blocking once again?

**Comment by yujuhong** [comment]
> It's currently part of Disruptive which is part of serial, which was probably disabled for this bug. Maybe we can just make serial blocking once again?

The test *was* in a blocking suite, but I guess the suite got removed from the blocking list because of legit test failures(?). If it's the policy to remove the suite temporarily, we should add it back once the investigation is completed. If serial suite simply runs too long (or has too much noise?), we can trim the tests and break it up to more suites. /cc @lavalamp

**Comment by bprashanth** [comment]
I believe serial was only failing post node/kubelet reboot when we disabled it

File: pkg/kubelet/network/kubenet/kubenet_linux.go [hunk 3]
```diff
@@ -249,8 +250,8 @@ func (plugin *kubenetNetworkPlugin) clearBridgeAddressesExcept(keep string) {
...
+			glog.V(2).Infof("Removing old address %s from %s", addr.IPNet.String(), BridgeName)
```
Comment (dims) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: Freehan, 
Was this a conscious change as well? log level from 5 to 2? (or just for debugging?)

**Comment by girishkalele** [comment]
@freehan @bprashanth @dchen1107 

LGTM ?

File: pkg/kubelet/network/kubenet/kubenet_linux.go (hunk 3 above)
Comment (dchen1107) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: The log is only generated one line upon the change when kubelet restart, right? In that case, it should be ok, and also can help further debugging.

**Comment by dchen1107** [comment]
LGTM

**Comment by gmarek** [comment]
@freehan Thanks a lot for fixing this!

**Comment by lavalamp** [comment]
> The test was in a blocking suite, but I guess the suite got removed from the blocking list because of legit test failures(?). If it's the policy to remove the suite temporarily, we should add it back once the investigation is completed. If serial suite simply runs too long (or has too much noise?), we can trim the tests and break it up to more suites. /cc @lavalamp

Yeah, I put serial tests as nonblocking because they break too often: having them as blocking or not didn't seem to change the speed with which issues were fixed, and significantly impacted the speed with which things were merged. Things you want to never ever break need to be tested in presubmit tests! Then we could turn it back on again.

**Comment by yujuhong** [comment]
> Yeah, I put serial tests as nonblocking because they break too often: having them as blocking or not didn't seem to change the speed with which issues were fixed, and significantly impacted the speed with which things were merged.

They break too often partially because a lot of tests in this suite test scenarios that don't happen in other suites (e.g., restarting components). 
Agreed that presubmit tests will be more ideal, but these test can not be run in parallel and will take too long to run for presubmit. Maybe we should try moving a subset of the tests to the presubmit tests.

That said, I think we'd still want a general policy for deciding whether a suite should be blocking or non-blocking.

File: pkg/kubelet/network/kubenet/kubenet_linux.go (hunk 3 above)
Comment (freehan) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: Yes. This only happens during kubelet initialization. The following line will take an important action. I feel like the log level should be V(2).

**Comment by lavalamp** [comment]
Yeah, the policy is being worked out in
https://docs.google.com/document/d/1jnGRrZymgnXoVUNnZVER_TOKH_Z11VvsKxlLxU1gg0A/edit#

short story:
* nothing in blocking post submits that's not tested (on a smaller scale if
needed) in presubmits
* If that means lots of separate presubmits, then that's what it means...

**Comment by yujuhong** [comment]
@lavalamp, thanks for the update!

**Pull request merged**